### PR TITLE
Add benchmark for libbeat diskqueue implementation

### PR DESCRIPTION
**Pull request by leehinman**
## What does this PR do?

Adds a benchmark test to libbeat disqueue implementation.  This benchmark produces, consumes & acks a million events to the queue.

## Why is it important?

This benchmark will allow us to see how changes to queue implementation affect performance.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
~~- [ ] I have made corresponding changes to the documentation~~
~~- [ ] I have made corresponding change to the default configuration files~~
- [x] I have added tests that prove my fix is effective or that my feature works
~~- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.~~

## How to test this PR locally

``` bash
go test -bench=. -benchtime 10x -timeout 60m
```

## Related issues

- Relates elastic/elastic-agent-shipper#33

### CODE REVIEW

File: libbeat/publisher/queue/diskqueue/benchmark_test.go [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+	}
+)
+
+//makeEvent creates a sample event, using a random message from msg above
+func makeEvent() publisher.Event {
+	return publisher.Event{
+		Content: beat.Event{
+			Timestamp: eventTime,
+			Fields: mapstr.M{
+				"message": msgs[rand.Intn(len(msgs))],
```
Comment (cmacknz) [inline libbeat/publisher/queue/diskqueue/benchmark_test.go:57]: Does using a random event make the benchmark results less reproducible? Is there a notable improvement vs just using a fixed sequence, or maybe the compromise of using the same seed in the random number generator for every run?

**Comment by cmacknz** [comment]
Can you summarize the results you get from this?

Also, does it run on every commit? It would be nice to get the output as report attached to each CI run, but that is maybe out of scope for this change.

File: libbeat/publisher/queue/diskqueue/benchmark_test.go [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+			},
+		},
+	}
+}
+
+//setup creates the disk queue, you must remove the directory it makes
```
Comment (fearful-symmetry) [inline libbeat/publisher/queue/diskqueue/benchmark_test.go:63]: Is there a reason we can't remove it after setup? If not, we might at least want to print the `dir` path so someone can manually clean it up without having to hunt around.

File: libbeat/publisher/queue/diskqueue/benchmark_test.go [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+	}
+	prod := q.Producer(queue.ProducerConfig{})
+	cons := q.Consumer()
+	return q, prod, cons
+}
+
+//clean closes the queue and deletes the directory
+func cleanup(q *diskQueue) {
+	q.Close()
+	os.RemoveAll(q.settings.directoryPath())
```
Comment (fearful-symmetry) [inline libbeat/publisher/queue/diskqueue/benchmark_test.go:84]: Wait, is this the temp path created in `setup` ? If so , might want to clarify the comment above the function.

**Review by faec — COMMENTED** [review]
My only question is similar to @cmacknz, how does this interact with CI? (Does it take long to run? Are the results surfaced anywhere when not running it manually?) If it doesn't have any expected "failure" outcome then should it be disabled on a default test run with a `+build` directive or something?

Also FYI, depending on which goes in first this may conflict with my pending PR that removes `queue.Consumer` from the queue API https://github.com/elastic/beats/pull/31502 -- it should be an easy merge though, the only difference is that `Get` is now a method on the queue itself so there's no need to create a consumer explicitly.

**Comment by leehinman** [comment]
> My only question is similar to @cmacknz, how does this interact with CI? (Does it take long to run? Are the results surfaced anywhere when not running it manually?) If it doesn't have any expected "failure" outcome then should it be disabled on a default test run with a `+build` directive or something?

`gotestsum` which runs all the go tests, does not run benchmarks.  So you need to run this by hand.  But I think that is reasonable.  It took ~7 minutes to run the benchmark 10 times on my Mac, and really you need to run 50 or 100 times to get enough data to start doing comparisons.  I don't think we can really justify that kind of energy consumption on every commit. 
> 
> Also FYI, depending on which goes in first this may conflict with my pending PR that removes `queue.Consumer` from the queue API #31502 -- it should be an easy merge though, the only difference is that `Get` is now a method on the queue itself so there's no need to create a consumer explicitly.

Thanks, should be an easy fix.

File: libbeat/publisher/queue/diskqueue/benchmark_test.go (hunk 3 above)
Comment (leehinman) [reply libbeat/publisher/queue/diskqueue/benchmark_test.go:84]: yep, I updated the comments.  let me know if that helps clear it up.  Basically the queue has to live on disk somewhere and we need to delete it when we are done.

File: libbeat/publisher/queue/diskqueue/benchmark_test.go (hunk 2 above)
Comment (leehinman) [reply libbeat/publisher/queue/diskqueue/benchmark_test.go:63]: updated comments

**Review by cmacknz — APPROVED** [review]
